securesystemslib/metadata.py

```diff
@@ -97,3 +99,79 @@ def pae(self) -> bytes:
             len(self.payload),
             self.payload,
         )
+
+
+class Parser:
+    """A Generic Payload Parser.
+
+    Attributes:
+        _supported_payload_types: list of payload types this parser can support.
+
+    Methods:
+        check_type(cls, payload_type: str):
+            Checks the parser support given payload type.
+
+        parse(envelope: Envelope):
+            Parse the envelope's payload.
+    """
+
+    __metaclass__ = abc.ABCMeta
+    _supported_payload_types: List[str] = []
+
+    @classmethod
+    def check_type(cls, payload_type: str) -> None:
+        """Checks the parser support given payload type.
+
+        Arguments:
+            payload_type: The payload_type to be checked.
+
+        Raises:
+            UnsupportedPayloadType: if payload_type is not a supported payload type.
+        """
+
+        if payload_type.lower() not in cls._supported_payload_types:
+            raise exceptions.UnsupportedPayloadType
+
+    @classmethod
+    @abc.abstractmethod
+    def parse(cls, envelope: Envelope) -> Any:
+        """Parse the envelope's payload.
+
+        Arguments:
+            envelope: The DSSE "Envelope" class instance.
+
+        Returns:
+            Returns the serialized body.
+        """
+        raise NotImplementedError  # pragma: no cover
+
+
+class JSONParser(Parser):
+    """A JSON/dict Parser for DSSE Envelope."""
+
+    _supported_payload_types: List[str] = ["json"]
```

Comment:
IIRC this isn't a legal payload type as per the DSSE spec.

Reply:
https://github.com/secure-systems-lab/dsse/blob/master/protocol.md

```
PAYLOAD_TYPE: Opaque, case-sensitive string that uniquely and unambiguously identifies how to interpret payload. This includes both the encoding (JSON, CBOR, etc.) as well as the meaning/schema. To prevent collisions, the value SHOULD be either:

    [Media Type](https://www.iana.org/assignments/media-types/), a.k.a. MIME type or Content Type
        Example: application/vnd.in-toto+json.
        IMPORTANT: This SHOULD be an application-specific type describing both encoding and schema, NOT a generic type like application/json. The problem with generic types is that two different applications could use the same encoding (e.g. JSON) but interpret the payload differently.
        SHOULD be lowercase.
    [URI](https://tools.ietf.org/html/rfc3986)
        Example: https://example.com/MyMessage/v1-json.
        SHOULD resolve to a human-readable description but MAY be unresolvable.
        SHOULD be case-normalized (section 6.2.2.1)
```